Feature branch/enumerable methods

bini-i opened the pull request:
The project is our own implementation of Ruby's enumerable methods and and contains the following methods
- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count?
- my_map
- my_inject  

and the project uses proc, blocks and yield to achieve the implementations

dibdas committed:
add enumerable

dibdas committed:
add my_each

dibdas committed:
add my_index

dibdas committed:
add my_select

bini-i committed:
Add my_all method

bini-i committed:
Add my_any method

bini-i committed:
Add my_none method

dibdas committed:
added my_count

dibdas committed:
added my_map

dibdas committed:
added my_inject

bini-i committed:
Add multiply_els method and update my_each and my_inject methods

bini-i committed:
Update my_map method

bini-i committed:
Update my_map method to take both a proc and block

bini-i committed:
Add git workflows

bini-i committed:
Update my_inject method

added an optional default parameter for the accumulator

bini-i committed:
Add README.md

bini-i committed:
Update README.md

bini-i committed:
Update README.md

bini-i committed:
Delete .rubocop.yml

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      yield ele
+    end
+    self
+  end
```
Is expected that:
#my_each returns an Enumerator if no block is given

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      yield self[indx], indx
+    end
+    self
+  end
```
Is expected that:
#my_each_with_index returns an enumerator if no block is given

#my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self

#my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

#my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      selected_elements.push(k) if yield k
+    end
+    selected_elements
+  end
```
Is expected that:
#my_select returns an enumerator if no block is given

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      return false unless yield ele
+    end
+    true
+  end
```
#my_all? when no block or argument is given returns true when none of the collection members are false or nil

#my_all? when no block or argument is given returns false when one of the collection members are false or nil

#my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer

Is expected that: 
#my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric

#my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class

#my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex

#my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex

#my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

#my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      return true if yield ele
+    end
+    false
+  end
```
Is expected that:

#my_any? when no block or argument is given returns true if at least one of the collection is not false or nil

#my_any? when no block or argument is given returns false if at least one of the collection is not true

#my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric

#my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer

#my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex

#my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex

#my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

#my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      return false if yield ele
+    end
+    true
+  end
```
Is expected that: 

#my_none? when no block or argument is given returns true only if none of the collection members is true

#my_none? when no block or argument is given returns false only if one of the collection members is true

#my_none? when a class is passed as an argument returns true if none of the collection is a member of such class

#my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric

#my_none? when a class is passed as an argument returns false if any of the collection is a member of such class

#my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex

#my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex

#my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

#my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      count += 1 if yield ele
+    end
+    count
+  end
```
Is expected that:

#my_count returns the number of items in enum through enumeration

#my_count returns the number of items in enum through enumeration::range

#my_count counts the number of items in enum that are equal to item if an argument is given

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      proc = param[0] # proc object
+    elsif param.length.zero? && block
+      proc = block # proc object
+    end
```
Is expected that :

#my_map returns an Enumerator if no block is given

kevcap commented on _enumerable.rb:
```diff
@@ -0,0 +1,84 @@
+      accumulator = yield accumulator, self[indx]
+    end
+    accumulator
+  end
```
Is expected that: 

#my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range

#my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block

#my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method

#my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range

#my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method

#my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range

kevcap reviewed (changes requested):
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:
 
## [REQUIRED CHANGES]
- [ ] Please note that your methods should work just as the original ruby methods, I've left some comments down below where you guys can read what is expected and by it, fix your methods.

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> <img  alt="gmail"  src="https://img.shields.io/badge/gmail-%20- green?logo=gmail&style=social"></a> [![Twitter Follow](https://img.shields.io/twitter/follow/kevinnnmn?style=social)](https://twitter.com/intent/follow?screen_name=kevinnnmn) [![](https://img.shields.io/badge/Slack-Kevin%20Alves-green?logo=slack&style=social)](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In real-world, you'll have a lot of changes requests for your real clients, it's good to get used to it.</sub>

bini-i committed:
Update my_each and my_each_with_index methods

to return enumerator when no block is passed
to iterate over ranges

bini-i committed:
Update my_all_method

to check when argument of type Regexp and class is passed
to return true when no block is passed and all elements are truthy
to return false when no block is passed and an element of falsy exists

bini-i committed:
Update my_map and my_inject methods

to return enumerator when no block is passed(for my_map).
when no block is given, to use a symbol as a named method to do an
operation on enum elements(for my_inject).

bini-i committed:
Update my_count method

to return number of matching elements when comparing value is given

bini-i committed:
Fix rubocop linter errors

bini-i committed:
Fix robocop linter errors

kubilaycaglayan reviewed (changes requested):
#### Hello Team 👋
#### Status: Changes Required ♻️

Great job so far! 👏👋
I see that you like to use various technologies to build your project! This is great 👏👍🎉. But please, when you are doing this make sure that you are giving decent instruction in your readme file about how to make the application work. 🙏😊

Let's cover a few details together to make your project better 😊⏬

### Readme

- [x] [REQUIRED] Add instructions about how to install `pry` if we need it to make the project work, please 😇🙏

### Naming Conventions

- [x] [OPTIONAL] What is the reason for the underscores at the beginning of your files? `_`
If you don't have a specific purpose for that, could you please delete it? 😊🙏

### Helper Methods

- [x] [REQUIRED] Dividing code into partials is great. But you are not calling this separate module in the enumerable? I should be able to use enumerable out of the box! Please add `require` and `include`.

### Enumerable Methods

#### my_all?
- [x] [REQUIRED] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] [REQUIRED] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
- [x] [REQUIRED] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [x] [REQUIRED] when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

#### my_none?
- [x] [REQUIRED] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] [REQUIRED] when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [x] [REQUIRED] when a class is passed as an argument returns false if any of the collection is a member of such class
- [x] [REQUIRED] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] [REQUIRED] when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

#### my_count
- [x] [REQUIRED] returns the number of items in enum through enumeration
- [x] [REQUIRED] returns the number of items in enum through enumeration::range

#### my_inject
- [x] [REQUIRED] when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
- [x] [REQUIRED] when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
- [x] [REQUIRED] when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
- [x] [REQUIRED] when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range

#### multiply_els
- [x] [REQUIRED] accepts an array as an argument and multiplies all the elements of the array together using #my_inject

## That's it, you can request another code review request after making the changes ♻️😇

With all my respect 🙏
> Kubilay Caglayan | Technical Support Engineer

bini-i committed:
Update my_all? and my_none? methods

In my_all? method
- for pattern other than regex and class made to return true only if all
elements match
In my_none? method
- for patter other than regex and class made to return
true only if there is no matching element
- fixed to check if all elements are of type given class

bini-i committed:
Update my_all? and my_none? methods

In my_all? method
- for pattern other than regex and class made to return true only if all
elements match
In my_none? method
- for patter other than regex and class made to return
true only if there is no matching element
- fixed to check if all elements are of type given class

bini-i committed:
Update README.md

bini-i committed:
Update README.md

bini-i committed:
Merge branch 'feature-branch/enumerable-methods' of github.com:dibdas/enumerable-methods into feature-branch/enumerable-methods

dibdas committed:
update my_count

bini-i committed:
Add multiply_els

Iterates through an array using my_inject and returns the product of
elements

bini-i committed:
Update README.md

maelfosso commented on enumerable.rb:
```diff
@@ -0,0 +1,143 @@
+    new_array
+  end
+
+  def my_inject(*param)
+    # when no block is given and symbol is provided as argument
+    Helper.no_block(self, param) unless block_given?
+
+    # when a block is specified
+    if param.length.positive?
+      # provides optional parameter for default value of accumulator
+      accumulator = param[0]
+      i = 0
+    elsif param.length.zero?
+      accumulator = first
+      i = 1
+    end
+    arr = *self # if a range is given, it splats it into an array
+    (i..(size - 1)).my_each do |indx|
+      accumulator = yield accumulator, arr[indx]
+    end
+    accumulator
+  end
+end
```
```rb
2.6.5 :013 > array
 => [7, 2, 0, 3, 1, 7, 1, 6, 4, 2, 1, 0, 8, 3, 1, 4, 0, 0, 8, 5, 7, 8, 0, 0, 6, 4, 4, 7, 1, 7, 5, 8, 2, 2, 4, 7, 6, 2, 7, 1, 5, 4, 4, 7, 4, 4, 4, 4, 6, 2, 0, 1, 4, 6, 0, 3, 3, 2, 8, 3, 4, 2, 2, 3, 0, 0, 4, 6, 1, 4, 5, 3, 6, 5, 0, 1, 2, 6, 3, 1, 5, 7, 5, 3, 0, 6, 4, 0, 8, 8, 3, 3, 8, 3, 7, 5, 3, 3, 7, 4] 
2.6.5 :014 > array.my_inject(:+)
Traceback (most recent call last):
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        3: from (irb):14
        2: from enumerable.rb:120:in `my_inject'
        1: from /home/maelfosso/Documents/Microverse/TSE/Ruby/enumerable-methods/helper_methods.rb:57:in `no_block'
NameError (undefined local variable or method `first' for Helper:Module)
2.6.5 :015 > array.inject(:+)
 => 375 
2.6.5 :016 > array.my_inject(20, :+)
Traceback (most recent call last):
       11: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
       10: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        9: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        6: from (irb):16
        5: from enumerable.rb:132:in `my_inject'
        4: from enumerable.rb:8:in `my_each'
        3: from enumerable.rb:8:in `times'
        2: from enumerable.rb:9:in `block in my_each'
        1: from enumerable.rb:133:in `block in my_inject'
LocalJumpError (no block given (yield))
2.6.5 :017 > array.inject(20, :+)
 => 395 
2.6.5 :018 > 
```

maelfosso replied:
```rb
2.6.5 :020 > range = Range.new(5, 50)
2.6.5 :021 > range.my_inject(:*)
Traceback (most recent call last):
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        3: from (irb):21
        2: from enumerable.rb:120:in `my_inject'
        1: from /home/maelfosso/Documents/Microverse/TSE/Ruby/enumerable-methods/helper_methods.rb:57:in `no_block'
NameError (undefined local variable or method `first' for Helper:Module)
2.6.5 :022 > range.inject(:*)
 => 1267253883404724085150525340252698701849068398706688000000000000 
2.6.5 :023 > range.my_inject(20, :*)
Traceback (most recent call last):
       11: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
       10: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        9: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        6: from (irb):23
        5: from enumerable.rb:132:in `my_inject'
        4: from enumerable.rb:8:in `my_each'
        3: from enumerable.rb:8:in `times'
        2: from enumerable.rb:9:in `block in my_each'
        1: from enumerable.rb:133:in `block in my_inject'
LocalJumpError (no block given (yield))
2.6.5 :024 > range.inject(20, :*)
 => 25345077668094481703010506805053974036981367974133760000000000000 
2.6.5 :025 > 
```

maelfosso reviewed (changes requested):
# Changes required 🔴 

Hello,

Thank you for implementing the suggested changes
However, the method `my_inject` is not working correctly.
Kindly read the comments left below, to know more about the bugs

Once fix, submit a new code review request
Thanks

Mael FOSSO

bini-i committed:
Fix my_inject method

fix call for first method to be on object of arr(array)

bini-i committed:
Fix my_inject method

fix call for first method to be on object arr

bini-i committed:
Merge branch 'feature-branch/enumerable-methods' of github.com:dibdas/enumerable-methods into feature-branch/enumerable-methods

euqueme commented on enumerable.rb:
```diff
@@ -0,0 +1,143 @@
+    false
+  end
+
+  def my_none?(*param)
+    return Helper.match_none(self, param[0]) if param.length.positive?
+
+    # returns true if a block is not given and truthy element is found
+    my_each { |ele| return false if ele && !block_given? }
+    # returns false if a block is not given and truthy element is not found
+    return true unless block_given?
+
+    my_each do |ele|
+      # returns false if an element that satsifys the condition is found
+      return false if yield ele
+    end
+    true
+  end
```
[OPTIONAL]

I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

```ruby
def my_none?(arg = nil, &block)
  !my_any?(arg, &block)
end
```

To make this work we have to use block.call(element) instead of yield(element) in my_any?

euqueme reviewed (changes requested):
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @dibdas and @bini-i :balloon: :lollipop: 

### Thank you for doing the changes requested by all the previous reviewers :medal_sports: :medal_sports: your project has come a long way and your enumerable methods are impeccable :sparkles: please feel proud of all your hard work :clap: :clap: :clap: there is only one change to make to your my_inject so let's get it working :construction: :construction_worker: :construction_worker: 

#### my_inject

- [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

- [x] Raises a "LocalJumpError" when no block or argument is given

```ruby
[1, 2, 3].my_inject == [1, 2, 3].inject => LocalJumpError
```
- my_inject is retrieving nil
![image](https://user-images.githubusercontent.com/52209067/96910753-2e864e80-145d-11eb-8fc0-22cb38dab1f2.png)

- Real inject retrieves LocalJumpError (correct answer)
![image](https://user-images.githubusercontent.com/52209067/96910874-537ac180-145d-11eb-9fdb-8af877e5f845.png)

### That's it, team! :raised_hands: 

### Please :pray: fix this :technologist: :technologist: and submit back for a code review when you are done :memo: 

### Happy Coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

bini-i committed:
Update my_inject method

to raise LocalJumpError when no block and no argument is provided

acushlakoncept reviewed (approved):
# Status: PROJECT APPROVED  🎉 🚀

#### Excellent Job Team!!! 👏 
I see you have successfully implemented all the changes requested by the previous TSE. 
And your methods have passed all tests.

I will go ahead and mark this as complete, most of your codebase is self-explanatory, take out the comments, and feel free to merge when you are ready. All the best with the next project.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)
